File: algorithms/2_medium/L/00522_longest_uncommon_subsequence_ii/solution.cpp
typedef ptrdiff_t isize;

class Solution {
    inline static bool is_sub(const string& a, const string& b) {
        const size_t a_len = a.size();
        const size_t b_len = b.size();

        if (a_len > b_len) {
            return false;
        }

        size_t a_idx = 0;
        size_t b_idx = 0;

        while (a_idx < a_len && b_idx < b_len) {
            if (a[a_idx] == b[b_idx]) {
                a_idx += 1;
            }
            b_idx += 1;
        }

        return a_idx == a_len;
    }
public:
    int findLUSlength(vector<string>& strs) {
        const size_t len = strs.size();
        assert(len >= 2);
        isize res = -1;
        for (size_t i = 0; i < len; i += 1) {
            for (size_t j = 0; j < len; j += 1) {
                if (i == j) { continue; }

                if (is_sub(strs[i], strs[j])) {
                    goto continue_main_loop;
                }
            }

            res = std::max(res, std::ssize(strs[i]));

            continue_main_loop:
                continue;
        }

        assert(res <= numeric_limits<int>::max());
        return static_cast<int>(res);
    }
};

File: algorithms/2_medium/L/00522_longest_uncommon_subsequence_ii/solution.rs
impl Solution {
    #[inline(always)]
    fn is_sub(a: &[u8], b: &[u8]) -> bool {
        if a.len() > b.len() { return false; }

        let (a_len, b_len) = (a.len(), b.len());
        let (mut a_idx, mut b_idx) = (0_usize, 0_usize);

        while a_idx < a_len && b_idx < b_len {
            if a[a_idx] == b[b_idx] {
                a_idx += 1;
            }
            b_idx += 1;
        }

        a_len == a_idx
    }

    pub fn find_lu_slength(strs: Vec<String>) -> i32 {
        let len = strs.len();
        assert!(len >= 2);
        let mut res = -1_isize;
        'main_loop: for i in 0..len {
            for j in 0..len {
                if i == j { continue; }

                if Self::is_sub(strs[i].as_bytes(), strs[j].as_bytes()) {
                    continue 'main_loop;
                }
            }

            res = res.max(strs[i].len() as isize);
        }

        res.try_into().unwrap()
    }
}
